### PR TITLE
Support escape expressions in @kwdef

### PR DESCRIPTION
**Pull request by ettersi**
`@kwdef` currently does not handle escaped type names and default values correctly. This makes it impossible to write a (correctly escaped) macro that internally uses `@kwdef`:

```julia
julia> module M
       macro define_struct(name)
           quote
               @kwdef struct $(esc(name)) field::Int end
           end
       end
       end;

julia> @macroexpand M.@define_struct(Foo)
ERROR: Invalid usage of @kwdef
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] var"@kwdef"(__source__::LineNumberNode, __module__::Module, expr::Any)
   @ Base ./util.jl:603
 [3] #macroexpand#66
   @ Base ./expr.jl:122 [inlined]
 [4] top-level scope
   @ REPL[5]:1
```

This PR patches `@kwdef` so various types of escaping are handled correctly. Specifically, all of the following now works (see new test case [here](https://github.com/ettersi/julia/blob/adf7e20378ae7d0c3cf24a2cc3c815e560913b78/test/misc.jl#L1290-L1332)):

```julia
module KwdefWithEsc
    const Int1 = Int
    const val1 = 42
    macro define_struct()
        quote
            @kwdef struct $(esc(:Struct))
                a
                b = val1
                c::Int1
                d::Int1 = val1


                $(esc(quote
                    e
                    f = val2
                    g::Int2
                    h::Int2 = val2
                end))


                $(esc(:(i = val2)))
                $(esc(:(j::Int2)))
                $(esc(:(k::Int2 = val2)))


                l::$(esc(:Int2))
                m::$(esc(:Int2)) = val1


                n = $(esc(:val2))
                o::Int1 = $(esc(:val2))


                $(esc(:p))
                $(esc(:q)) = val1
                $(esc(:s))::Int1
                $(esc(:t))::Int1 = val1
            end
        end
    end
end


module KwdefWithEsc_TestModule
    using ..KwdefWithEsc
    const Int2 = Int
    const val2 = 42
    KwdefWithEsc.@define_struct()
end
```

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
Seems good to me. Some implementation thoughts though to make it even better:

If possible, using a `rewrap_esc` implementation which takes the old expr to mimic instead of just a count, might be better than `wrap_esc`, since it would permit it to correctly handle `escape`'s dual: the `hygienic-scope` expr.

Note that there is also an existing implementation of `unblock`/`unescape` in `base/docs/Docs.jl` (because `@docs` has to deal with this exact same problem) which has some extra features this may be missing. We might want to move that functionality to `expr.jl` or `meta.jl` so that it can be shared with this code?

E.g. here's where this same issue is handled for a docstring:
https://github.com/JuliaLang/julia/blob/d24316ac6465302aa32d9616ab731499c3734bce/base/docs/Docs.jl#L368-L381

**Comment by vtjnash** [comment]
Aside: for what it is worth, people usually probably got around this bug in the past by moving the `esc` call like so:
```
       macro define_struct(name)
           quote
               $(esc(:($(@__MODULE__).@kwdef struct $name; field::Int; end)))
           end
       end
```

File: base/meta.jl [hunk 1]
```diff
@@ -461,4 +464,45 @@ end
...
+function unescape(@nospecialize ex)
+    ex = unblock(ex)
+    while isexpr(ex, :escape) || isexpr(ex, :var"hygienic-scope")
+       ex = unblock(ex.args[1])
+    end
+    return ex
+end
+
+"""
+    uncurly(expr)
```
Comment (inkydragon) [inline base/meta.jl:502]:
```suggestion
    Meta.uncurly(expr)
```

File: base/meta.jl [hunk 2]
```diff
@@ -16,7 +16,10 @@ export quot,
        ispostfixoperator,
        replace_sourceloc!,
        show_sexpr,
-       @dump
+       @dump,
+       unblock,
+       unescape,
+       uncurly
```
Comment (inkydragon) [inline base/meta.jl:22]: Do we want to export these functions now?
Maybe keep the status quo and let them remain as internal functions.

```suggestion
       @dump
```

Comment (ettersi) [reply base/meta.jl:22]: Will commit your suggestions as soon as the tests have finished 🙏

**Comment by ettersi** [comment]
@vtjnash I believe I have incorporated all the changes that you requested. I hope it's okay that I rearranged the code quite a bit while doing so. Also, the test failures seem to be completely unrelated to the changes in this PR. Any idea what is going on there?

**Comment by vtjnash** [comment]
This rearrangement looks great! I think it looked like there was a temporary network failure at our CI host. Hopefully this current build will be unaffected by it.

**Pull request merged**